llvm/lib/Target/WebAssembly/WebAssemblyMemorySafety.cpp

```diff
@@ -198,6 +191,46 @@ class WebAssemblyMemorySafety : public FunctionPass {
   }
 };
 
+bool isAllocKind(Attribute Attr, AllocFnKind Kind) {
+  if (!Attr.hasAttribute(Attribute::AllocKind))
+    return false;
+
+  return (Attr.getAllocKind() & Kind) != AllocFnKind::Unknown;
+}
+
+// A flattened (each option contains no more "hidden" options through a bitmap)
+// version of AllocKind. This contains all AllocKind's that C supports, so if we
+// ever want to support more languages, we would have to extend this.
+enum class FlattenedAllocKind {
+  // Alloc + Uninitialized
+  Malloc,
+  // Alloc + Zeroed
+  Calloc,
+  // Realloc
+  Realloc,
+  // Free
+  Free,
+  // We don't handle/support these kinds of AllocKind at all
+  Unhandled,
+};
+
+FlattenedAllocKind flattenAllocKind(Attribute Attr) {
+  // TODO: do we also need to handle aligned? Or do we not need that at all since we always align ourselves?
```

Comment:
Just pointing you to some open questions I had.

Reply:
I guess yes, as a user might request alignment > 16, e.g. `aligned_alloc(32, size)`. We should align to the requested 32 bytes instead of 16 in this case. I am not totally sure if this is needed for our benchmarks, though.